mgr/src/server/plugins/volume_expand.cr

```diff
@@ -0,0 +1,274 @@
+  if !resp.ok
+    halt(env, status_code: 400, response: node_errors("Not all participant nodes are reachable", resp.node_responses).to_json)
+  end
+
+  puts "after ping"
+
+  # If User specified the Port in the request then validate if
+  # the port is already used.
+  req.distribute_groups.each do |dist_grp|
+    dist_grp.storage_units.each do |storage_unit|
+      next if storage_unit.port == 0
+
+      unless Datastore.port_available?(pool.not_nil!.id, storage_unit.node.id, storage_unit.port)
+        # TODO: Move this halt out of the loop
+        halt(env, status_code: 400, response: ({"error": "Port is already used(#{storage_unit.node.name}:#{storage_unit.port})"}).to_json)
+      end
+    end
+  end
+
+  # Local Validations
+  resp = dispatch_action(
+    ACTION_VALIDATE_VOLUME_CREATE,
+    pool_name,
+    nodes,
+    req.to_json
+  )
+
+  if !resp.ok
+    halt(env, status_code: 400, response: node_errors("Invalid Volume expand request", resp.node_responses).to_json)
+  end
+
+  # Update Free Ports and reserve it
+  # Also generate Brick ID
+  # TODO: Update Brick type
+  no_port_available = false
+  no_port_storage_node = ""
+  req.distribute_groups.each do |dist_grp|
+    dist_grp.storage_units.each do |storage_unit|
+      storage_unit.id = UUID.random.to_s
+      # Request doesn't contain the Port, find a free port
+      if storage_unit.port == 0
+        free_port = Datastore.free_port(pool.not_nil!.id, storage_unit.node.id)
+        storage_unit.port = free_port unless free_port.nil?
+      end
+
+      # No free port found
+      if storage_unit.port == 0
+        # Using halt directly from here will not work. Since it adds `next`
+        # and it just breaks the Storage_unit loop
+        no_port_available = true
+        no_port_storage_node = storage_unit.node.name
+        break
+      end
+
+      Datastore.reserve_port(pool.not_nil!.id, storage_unit.node.id, storage_unit.port)
+    end
+
+    break if no_port_available
+  end
+
+  if no_port_available
+    halt(env, status_code: 400, response: ({"error": "No free Port available in #{no_port_storage_node}"}).to_json)
+  end
+
+  # Generate Services and Volfiles if Volume to be started
+  services, volfiles = services_and_volfiles(req)
+
+  action = ACTION_VOLUME_CREATE
+  req.state = "Started"
+  if req.no_start
+    action = ACTION_VOLUME_CREATE_STOPPED
+    req.state = "Created"
+  end
+
+  # Volume create action {req, services, volfiles}
+  resp = dispatch_action(
+    action,
+    pool_name,
+    nodes,
+    {services, volfiles, req}.to_json
+  )
+
+  if !resp.ok
+    halt(env, status_code: 400, response: node_errors("Failed to expand Volume", resp.node_responses).to_json)
+  end
+
+  # Save Services details... Causing exception of unique constraint failed
+  # services.each do |node_id, svcs|
+  #   svcs.each do |svc|
+  #     # Enable each Services
+  #     Datastore.enable_service(pool.not_nil!.id, node_id, svc)
```

Comment:
This fails to add services to DB since node_id and pool_id will be same if we add a new storage_unit in the same brick. Throws `Unique constraint failed`.

Reply:
This is maybe beacuse shd services are being added to db for already existing node.id